client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -168,24 +170,35 @@ private String updateGroupStateIfNeeded() throws ReaderNotInReaderGroupException
             groupState.checkpoint(atCheckpoint, getPosition());
             releaseSegmentsIfNeeded();
         }
-        atCheckpoint = groupState.getCheckpoint();
-        if (atCheckpoint != null) {
-            log.info("{} at checkpoint {}", this, atCheckpoint);
-            if (groupState.isCheckpointSilent(atCheckpoint)) {
+        String checkpoint = groupState.getCheckpoint();
+        if (checkpoint != null) {
+            log.info("{} at checkpoint {}", this, checkpoint);
+            if (groupState.isCheckpointSilent(checkpoint)) {
                 // Checkpoint the reader immediately with the current position. Checkpoint Event is not generated.
-                groupState.checkpoint(atCheckpoint, getPosition());
-                atCheckpoint = null;
+                groupState.checkpoint(checkpoint, getPosition());
+                if (atCheckpoint != null) {
+                    //In case the silent checkpoint held up releasing segments
+                    releaseSegmentsIfNeeded();
```

Comment:
if `atCheckpoint` is not null then releaseSegmentIfNeeded() will be invoked at line 171.  Did not understand the need to invoke it again.

Reply:
The releaseSegmentIfNeeded() at 171 will not release any segments if a checkpoint is pending, so that the checkpoint can be performed first. In the case that this is a normal checkpoint this is fine, but in the case that it is a silent checkpoint this is not desirable. 

I understand that that is confusing. So I am changing this to handled internally to relaeasSegment(). That way these sort of operations are not affected by silent checkpoints.